### PR TITLE
GRDB for REST models

### PR DESCRIPTION
**Pull request by ualch9**
This is the previous PR (#684), but targeting feature branch `ualch9/persistence` and without `MetaCodable`. GRDB's Codable support was fighting with the code generated by MetaCodable, so I decided to drop it.

- Added new `OBAKitCoreTests` target for testing against `OBAKitCore.framework` only.
    - OBAKitTests depends on `OBAKit.framework` and `App.app`.
    - OBAKitCoreTests links against only `OBAKitCore.framework`, so I can focus on Codable/GRDB without fighting the compiler or working without tests. Where appropriate, I moved some tests were moved from OBAKitTests → OBAKitCoreTests.
    - For now, fixture files are included in both OBAKitTests and OBAKitCoreTests target.

- GRDB and REST implementations are contained in their own files.
    - REST models are in `OBAKitCore/Models/REST`. 
    - GRDB implementations are Swift extensions in `OBAKitCore/Persistence/Models`. In general, GRDB models (such as many-to-many relations) are not exposed outside of OBAKitCore.
    - With few exceptions, REST and GRDB use the same model.

Each model defines their own tables, via `protocol DatabaseTableCreator.createTable(in:)` that is called by `PersistenceManager` when initially creating the database.

When storing on-disk, each database is stored per-region: `{app_container}/Application Support/regions/{region_id}/onebusaway.sqlite`

Screenshot 2023-10-26 at 12 12 26 PM

Undecided:
- What should be persisted (stops only? or everything?). As a first release, the database could operate only in-memory.
- Need to include some sort of expiration date. AFAIK, API responses do not include an `expires` header.
- How should bookmarks be stored so they can sync via iCloud.
- Objective-C support? Is this important enough that we should write wrappers for the new models?

### CODE REVIEW

**Comment by aaronbrethorst** [comment]
> What should be persisted (stops only? or everything?). As a first release, the database could operate only in-memory.
> Need to include some sort of expiration date. AFAIK, API responses do not include an expires header.

Only in memory sounds like a great first step. That'll help us deal with this problem you outline in point 2 (above).

> How should bookmarks be stored so they can sync via iCloud.

Let's defer this just for a little bit. I think getting Sqlite up and running in a way we're comfortable with supporting is enough of a bite at this apple for now.

> Objective-C support? Is this important enough that we should write wrappers for the new models?

I don't think so, no. I think we can effectively ignore Obj-C from here on out.

File: OBAKitCore/Bookmarks/TripBookmarkKey.swift [hunk 1]
```diff
@@ -28,7 +28,8 @@ public struct TripBookmarkKey: Hashable, Equatable {
     }
 
     public init(arrivalDeparture: ArrivalDeparture) {
```
Comment (aaronbrethorst) [inline OBAKitCore/Bookmarks/TripBookmarkKey.swift:30]: it seems confusing to me to keep this `init` method around—only to have it generate a `fatalError()` if called. Seems like a major footgun. Please delete unless I'm missing something important.

File: OBAKitCore/Models/REST/References/SituationREST.swift [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+/// Model for decoding Situation objects in the format of the REST API.
+/// To insert a Situation into an SQLite database, use ``SituationREST.insert(into:)`` helper, which
+/// creates a ``Situation`` struct and creates the necessary relationships.
+///
+/// In general, ``SituationREST`` is only for decoding from the OBA API. Use ``Situation`` for all other cases.
+struct SituationREST: Decodable, Identifiable {
+    // SituationREST is a separate data model from Situation because
+    // some relationships are better represented as relation tables 
+    // in the database for a nicer SQLite querying experience.
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/References/SituationREST.swift:20]: 👍

File: OBAKitCore/Models/REST/References/Stop.swift [hunk 3]
```diff
@@ -89,13 +89,22 @@ public class Stop: NSObject, Identifiable, Codable, HasReferences {
...
     /// Identifies whether this stop represents a stop, station, or station entrance.
     public let locationType: StopLocationType
 
     /// A human-readable name for this stop.
     public let name: String
 
+    /// A list of route IDs served by this stop.
+    ///
+    /// This value is only decoded/encoded via Codable. When retrieving from GRDB, this value will be empty.
+    let routeIDs: [RouteID]
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/References/Stop.swift:101]: are there circumstances where this being empty when retrieved from GRDB is going to be problematic?

File: OBAKitCore/Models/REST/References/Stop.swift [hunk 4]
```diff
@@ -106,45 +115,34 @@ public class Stop: NSObject, Identifiable, Codable, HasReferences {
...
-    public let routeIDs: [String]
-
-    /// A list of `Route`s served by this stop.
-    public var routes: [Route]!
-
     /// All unique route types at this Stop.
-    public lazy var routeTypes: Set<Route.RouteType> = {
-        return Set(routes.map { $0.routeType })
-    }()
+    public var routeTypes: Set<Route.RouteType> {
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/References/Stop.swift:119]: delete this if it's going to remain unimplemented. 🦶 🔫

File: OBAKitCore/Models/REST/References/Stop.swift [hunk 5]
```diff
@@ -106,45 +115,34 @@ public class Stop: NSObject, Identifiable, Codable, HasReferences {
...
-        // swiftlint:disable for_where
-        for t in priorities {
-            if routeTypes.contains(t) {
-                return t
-            }
-        }
-        // swiftlint:enable for_where
-
-        return .unknown
+        fatalError("\(#function) unimplemented.")
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/References/Stop.swift:126]: 🦶 🔫

File: OBAKitCore/Models/REST/References/Stop.swift [hunk 6]
```diff
@@ -106,45 +115,34 @@ public class Stop: NSObject, Identifiable, Codable, HasReferences {
...
-    public private(set) var regionIdentifier: Int?
-
-    private enum CodingKeys: String, CodingKey {
+    internal enum CodingKeys: String, CodingKey {
         case code
         case direction
         case id
-        case lat
-        case lon
+        case latitude = "lat"
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/References/Stop.swift:144]: 👍

File: OBAKitCore/Models/REST/References/Trip.swift [hunk 7]
```diff
@@ -76,68 +71,28 @@ public class Trip: NSObject, Identifiable, Decodable, HasReferences {
...
-        hasher.combine(id)
-        hasher.combine(regionIdentifier)
-        hasher.combine(routeID)
-        hasher.combine(routeShortName)
-        hasher.combine(serviceID)
-        hasher.combine(shapeID)
-        hasher.combine(shortName)
-        hasher.combine(timeZone)
-        return hasher.finalize()
+        fatalError("\(#function) unimplemented")
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/References/Trip.swift:92]: 🦶 🔫

File: OBAKitCore/Models/REST/StopArrivals.swift [hunk 8]
```diff
@@ -18,56 +18,29 @@ public class StopArrivals: NSObject, Identifiable, Decodable, HasReferences {
...
     /// Returns this model's list of service alerts, if any exist. If this model does not have any, then it returns a flattened list of its `ArrivalDepartures` objects' service alerts.
-    public var serviceAlerts: [ServiceAlert] {
-        if _serviceAlerts.count > 0 {
-            return _serviceAlerts
-        }
-        else {
-            return arrivalsAndDepartures.flatMap { $0.serviceAlerts }
-        }
-    }
+//    public var serviceAlerts: [ServiceAlert] {
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/StopArrivals.swift:27]: delete me

File: OBAKitCore/Models/REST/TripStatus.swift [hunk 9]
```diff
@@ -194,83 +189,26 @@ public class TripStatus: NSObject, Identifiable, Decodable, HasReferences {
...
+        try container.encode(closestStopID, forKey: .closestStopID)
+        try container.encode(closestStopTimeOffset, forKey: .closestStopTimeOffset)
+        try container.encode(distanceAlongTrip, forKey: .distanceAlongTrip)
+        try container.encodeIfPresent(frequency, forKey: .frequency)
+        try container.encode(lastKnownDistanceAlongTrip, forKey: .lastKnownDistanceAlongTrip)
+        try container.encode(lastKnownOrientation, forKey: .lastKnownOrientation)
+        try container.encode(lastLocationUpdateTime, forKey: .lastLocationUpdateTime)
+
+        try container.encodeIfPresent(lastKnownLocation, forKey: .lastKnownLocation)
+//        try container.encodeIfPresent(lastKnownLocation?.asOBALocationModel(), forKey:.lastKnownLocation)
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/TripStatus.swift:206]: deleteme

File: OBAKitCore/Models/REST/VehicleStatus.swift [hunk 10]
```diff
@@ -72,13 +67,8 @@ public class VehicleStatus: NSObject, Identifiable, Decodable, HasReferences {
 
         phase = try container.decode(String.self, forKey: .phase)
         status = try container.decode(String.self, forKey: .status)
-        location = try? CLLocation(container: container, key: .location)
+        location = try container.decode(LocationModel.self, forKey: .location)
+//        location = try? CLLocation(container: container, key: .location)
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/REST/VehicleStatus.swift:71]: deleteme

File: OBAKitCore/Models/UserData/Bookmarks/Bookmark.swift [hunk 11]
```diff
@@ -60,7 +60,8 @@ import Foundation
     // MARK: - Init
 
     public convenience init(name: String, regionIdentifier: Int, arrivalDeparture: ArrivalDeparture) {
-        self.init(name: name, regionIdentifier: regionIdentifier, arrivalDeparture: arrivalDeparture, stop: arrivalDeparture.stop)
+        fatalError("\(#function) unimplemented")
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/UserData/Bookmarks/Bookmark.swift:63]: 🦶 🔫

File: OBAKitCore/Models/UserData/UserDataStore.swift [hunk 12]
```diff
@@ -108,26 +108,26 @@ public protocol UserDataStore: NSObjectProtocol {
 
     /// Find recent stops that match `searchText`
     /// - Parameter searchText: The search string
-    func findRecentStops(matching searchText: String) -> [Stop]
+//    func findRecentStops(matching searchText: String) -> [Stop]
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/UserData/UserDataStore.swift:111]: please delete this and the other commented-out methods in the protocol

File: OBAKitCore/Models/ViewModels/TransitAlertViewModel.swift [hunk 13]
```diff
@@ -21,27 +21,27 @@ extension AgencyAlert: TransitAlertViewModel {
...
-        var body = ""
-
-        if let situationDescription = situationDescription {
-            body = "\(situationDescription.value)\r\n\r\n"
-        }
-
-        if affectedAgencies.count > 0 {
-            body = body.appending(affectedAgencies.map { $0.name }.sorted().joined(separator: ", "))
-        }
+//extension ServiceAlert: TransitAlertViewModel {
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/ViewModels/TransitAlertViewModel.swift:24]: delete

File: OBAKitCore/Models/ViewModels/TripConvertible.swift [hunk 14]
```diff
@@ -20,10 +20,11 @@ public class TripConvertible: NSObject {
     }
 
     public init?(vehicleStatus: VehicleStatus) {
-        guard vehicleStatus.trip != nil else {
-            return nil
-        }
-        self.vehicleStatus = vehicleStatus
+        fatalError("\(#function) unimplemented.")
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/ViewModels/TripConvertible.swift:23]: 🦶 🔫

File: OBAKitCore/Models/ViewModels/TripConvertible.swift [hunk 15]
```diff
@@ -35,7 +36,8 @@ public class TripConvertible: NSObject {
     }
 
     public var trip: Trip {
-        return arrivalDeparture?.trip ?? vehicleStatus!.trip!
+        fatalError("\(#function) unimplemented.")
```
Comment (aaronbrethorst) [inline OBAKitCore/Models/ViewModels/TripConvertible.swift:39]: 🦶 🔫

File: OBAKitCore/Orchestration/CoreApplication.swift [hunk 16]
```diff
@@ -200,7 +200,8 @@ open class CoreApplication: NSObject,
 
     public func migrate(recentStop: Stop) {
         guard let currentRegion = currentRegion else { return }
-        userDataStore.addRecentStop(recentStop, region: currentRegion)
+        fatalError("\(#function) unimplemented.")
```
Comment (aaronbrethorst) [inline OBAKitCore/Orchestration/CoreApplication.swift:203]: 🦶 🔫

File: OBAKitCore/Services/Persistence/PersistenceService.swift [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+            database = try DatabaseQueue(path: databaseURL.path)
+        case .memory:
+            logger.info("Opening SQLite in memory.")
+            database = try DatabaseQueue()
+        }
+
+        // Migration registration
+        var migrator = DatabaseMigrator()
+        migrator.registerMigration("createUsingTableCreators") { db in
+            let logger = os.Logger(subsystem: "org.onebusaway.iphone", category: "PersistenceService-DatabaseMigrator")
```
Comment (aaronbrethorst) [inline OBAKitCore/Services/Persistence/PersistenceService.swift:73]: `subsystem`'s value probably should not be hardcoded here.

File: OBAKitCore/Services/Persistence/PersistenceService.swift [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+
+            // Additional tables are usually for relationships. These should be created last, since SQLite needs to reference other tables.
+            for additionalCreator in additionalTableCreators {
+                logger.trace("Creating additional table: \(String(describing: additionalCreator))")
+                try additionalCreator.createTable(in: db)
+            }
+        }
+
+        #if DEBUG
+        migrator.eraseDatabaseOnSchemaChange = true
```
Comment (aaronbrethorst) [inline OBAKitCore/Services/Persistence/PersistenceService.swift:90]: I'd actually be ok with this always being true.

File: OBAKitCore/Utilities/Formatters.swift [hunk 19]
```diff
@@ -485,10 +485,10 @@ public class Formatters: NSObject {
         if let direction = directionString(stop.direction) {
             parts.append(direction)
         }
-
-        if let routes = Formatters.formattedRoutes(stop.routes) {
-            parts.append(routes)
-        }
+        fatalError("\(#function) unimplemented")
```
Comment (aaronbrethorst) [inline OBAKitCore/Utilities/Formatters.swift:488]: 🦶 🔫

File: OBAKitCore/Utilities/Formatters.swift [hunk 20]
```diff
@@ -631,11 +631,12 @@ public class Formatters: NSObject {
     /// - Parameter routes: An array of `Route`s from which the string will be generated.
     /// - Returns: A human-readable list of the passed-in `Route`s.
     public class func formattedAgenciesForRoutes(_ routes: [Route]) -> String {
-        return routes
-            .compactMap { $0.agency.name }
-            .uniqued
-            .sorted()
-            .joined(separator: ", ")
+        fatalError("\(#function) unimplemented")
```
Comment (aaronbrethorst) [inline OBAKitCore/Utilities/Formatters.swift:634]: 🦶 🔫

**Review by aaronbrethorst — CHANGES_REQUESTED** [review]
this looks like it's going to be a great step forward for OBAKit! I have some minor nits that I've noted throughout. Thanks for spearheading this, Alan!

File: OBAKitCore/Models/REST/References/Stop.swift (hunk 3 above)
Comment (ualch9) [reply OBAKitCore/Models/REST/References/Stop.swift:101]: GRDB will return an empty array if I fetch for `ids: []`. Also, I actually do insert the `routeIDs` into GRDB, so I removed that comment.

File: OBAKitCore/Bookmarks/TripBookmarkKey.swift (hunk 1 above)
Comment (ualch9) [reply OBAKitCore/Bookmarks/TripBookmarkKey.swift:30]: I wrote these fatalErrors in areas that used to rely on the `loadReferences` infrastructure, but I don't yet fully understand how OBAKit uses them. It exists to temporarily satisfy the compiler until I go back in a later PR to address them. I updated the error string to include `"PR-686 error"` to grep later.

File: OBAKitCore/Services/Persistence/PersistenceService.swift (hunk 17 above)
Comment (ualch9) [reply OBAKitCore/Services/Persistence/PersistenceService.swift:73]: Changed to whitelabel-friendly `OBAKitCore`.

**Review by aaronbrethorst — APPROVED** [review]
thanks for addressing my feedback, @ualch9! looks great!

**Pull request merged**